src/mam4xx/convproc.hpp

```diff
@@ -752,7 +756,192 @@ void ma_precpevap_convproc(
+    const int km1x = haero::max(kk - 1, 0);
+    const Real fa_u_dp = fa_u[kk] * dpdry_i[kk];
+    for (int icnst = 1; icnst < ConvProc::pcnst_extd; ++icnst) {
+      if (doconvproc_extd[icnst]) {
+        // compute fluxes as in convtran, and also source/sink terms
+        // (version 3 limit fluxes outside convection to mass in appropriate
+        // layer (these limiters are probably only safe for positive definite
+        // quantitities (it assumes that mu and md already satify a courant
+        // number limit of 1)
+        Real fluxin = 0, fluxout = 0;
+        if (iflux_method != 2) {
+          fluxin =
+              mu_i[kp1] * conu(kp1, icnst) +
+              mu_i[kk] * haero::min(chat(kk, icnst), gath(km1x, icnst)) -
+              (md_i[kk] * cond(kk, icnst) +
+               md_i[kp1] * haero::min(chat(kp1, icnst), gath(kp1x, icnst)));
+          fluxout = mu_i[kk] * conu(kk, icnst) +
+                    mu_i[kp1] * haero::min(chat(kp1, icnst), gath(kk, icnst)) -
+                    (md_i[kp1] * cond(kp1, icnst) +
+                     md_i[kk] * haero::min(chat(kk, icnst), gath(kk, icnst)));
+        } else {
+          // new method -- simple upstream method for the env subsidence
+          // tmpa = net env mass flux (positive up) at top of layer k
+          fluxin = mu_i[kp1] * conu(kp1, icnst) - md_i[kk] * cond(kk, icnst);
+          fluxout = mu_i[kk] * conu(kk, icnst) - md_i[kp1] * cond(kp1, icnst);
+          Real tmpa = -(mu_i[kk] + md_i[kk]);
+          if (tmpa <= 0.0) {
+            fluxin -= tmpa * gath(km1x, icnst);
+          } else {
+            fluxout += tmpa * gath(kk, icnst);
+          }
+          // tmpa = net env mass flux (positive up) at base of layer k
+          tmpa = -(mu_i[kp1] + md_i[kp1]);
+          if (tmpa >= 0.0) {
+            fluxin += tmpa * gath(kp1x, icnst);
+          } else {
+            fluxout -= tmpa * gath(kk, icnst);
+          }
+        }
+        //  net flux [kg/kg/s * mb]
+        const Real netflux = fluxin - fluxout;
+
+        // note for C++ refactoring:
+        // I was trying to separate dconudt_activa and dconudt_wetdep out
+        // into a subroutine, but for some reason it doesn't give consistent
+        // dcondt values. have to leave them here.   Shuaiqi Tang, 2022
+        const Real netsrce =
+            fa_u_dp * (dconudt_activa(kk, icnst) + dconudt_wetdep(kk, icnst));
+        dcondt(kk, icnst) = (netflux + netsrce) / dpdry_i[kk];
+      }
+    }
+  }
+}
+// =========================================================================================
+// TODO: compute_downdraft_mixing_ratio uses multiple levels for computation but
+// ONLY sets a SINGLE level on output for the loop over ktop to kbot.  So, it
+// should be possible to input kk and then call in parallel from ktop to kbot.
+// It is a bit confusing that the level set is kk+1 so it would be better to
+// rewrite as kkp1=>kk and kk=>kk-1 and iterate from ktop+1 to kbot inclusive.
```

Comment:
Starting with the bottom functions, it is sometimes not obvious how the interface might be changed to allow for parallel over levels once integrated into higher level functions.  But I am keeping this in mind and expect to eventually have Kokkos parallel loops over teams for many but maybe not all of the sub-functions.  I figure it will not be too hard to tweak working and tested C++ code at that time.